### PR TITLE
Do not allow retention rules to be null

### PR DESCRIPTION
**Pull request by kfaraz**
### Changes

- Give a proper error message when retention rules are passed as null/empty to the `POST coordinator/v1/rules` API
- Empty rules are allowed at the datasource level but not at the cluster level
- Add validation to ensure that `druid.manager.rules.defaultRule` is always set correctly
- Minor style refactors

### Changed classes
- Update `RulesResource`: Handle `IllegalArgumentException` and send back appropriate response with error message
- Update `SqlMetadataRuleManager`
  - Throw an exception when rules are null/empty
  - Minor style refactor using lambdas
- Refactor `AuditInfo`: Clean equals and hashCode methods
- Add `AuditInfoTest`
- Refactor `SqlMetadataRuleManagerTest`

### Screenshots

empty_cluster_rules
null_rules

#### Release note
- Rules defined at the datasource level cannot be null
- Rules defined at the cluster level cannot be null or empty

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataRuleManager.java [hunk 1]
```diff
@@ -362,6 +325,12 @@ public List<Rule> getRulesWithDefault(final String dataSource)
   @Override
   public boolean overrideRule(final String dataSource, final List<Rule> newRules, final AuditInfo auditInfo)
   {
+    if (newRules == null) {
+      throw new IAE("Rules cannot be null.");
+    } else if (newRules.isEmpty() && config.getDefaultRule().equals(dataSource)) {
```
Comment (AmatyaAvadhanula) [inline server/src/main/java/org/apache/druid/metadata/SQLMetadataRuleManager.java:330]: Could you please explain this condition: `config.getDefaultRule().equals(dataSource)`?

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataRuleManager.java:330]: The `defaultRule` corresponds to the cluster level default. I will add a comment here explaining it.

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataRuleManager.java:330]: Oh wait, I think the IAE thrown in the next line explains it.

File: server/src/test/java/org/apache/druid/metadata/SQLMetadataRuleManagerTest.java [hunk 2]
```diff
@@ -276,36 +277,29 @@ public void testRemoveRulesOlderThanWithNonExistenceDatasourceAndNewerThanTimest
...
+            ImmutableMap.of(DruidServer.DEFAULT_TIER, DruidServer.DEFAULT_NUM_REPLICANTS)
         )
     );
-    AuditInfo auditInfo = new AuditInfo("test_author", "test_comment", "127.0.0.1");
-    ruleManager.overrideRule(
-        "test_dataSource",
-        rules,
-        auditInfo
-    );
+    ruleManager.overrideRule(DATASOURCE, rules, createAuditInfo("udpate rules"));
```
Comment (AmatyaAvadhanula) [inline server/src/test/java/org/apache/druid/metadata/SQLMetadataRuleManagerTest.java:292]: nit: udpate -> update

File: server/src/test/java/org/apache/druid/metadata/SQLMetadataRuleManagerTest.java [hunk 3]
```diff
@@ -250,23 +257,17 @@ public void testRemoveRulesOlderThanWithNonExistenceDatasourceAndNewerThanTimest
...
         )
-        )
-    );
-    AuditInfo auditInfo = new AuditInfo("test_author", "test_comment", "127.0.0.1");
-    ruleManager.overrideRule(
-        "test_dataSource",
-        rules,
-        auditInfo
     );
+    ruleManager.overrideRule(DATASOURCE, rules, createAuditInfo("udpate rules"));
```
Comment (AmatyaAvadhanula) [inline server/src/test/java/org/apache/druid/metadata/SQLMetadataRuleManagerTest.java:264]: nit: udpate -> update

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataRuleManager.java (hunk 1 above)
Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataRuleManager.java:330]: I see, so the defaultRule in the config corresponds to the datasource's name in the default load rules.

**Review by AmatyaAvadhanula — APPROVED** [review]

File: server/src/main/java/org/apache/druid/metadata/SQLMetadataRuleManager.java (hunk 1 above)
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/metadata/SQLMetadataRuleManager.java:330]: Added a javadoc for `getDefaultRule`.

**Pull request merged**